pyk/src/pyk/cterm/cterm.py

```diff
@@ -253,6 +258,48 @@ def remove_useless_constraints(self, keep_vars: Iterable[str] = ()) -> CTerm:
         return CTerm(self.config, new_constraints)
 
 
+def merge_cterms(t1: CTerm, t2: CTerm) -> CTerm | None:
+    """Return a `CTerm` which is the merge of the two input `CTerm` instances.
+
+    Args:
+        t1: First `CTerm` to merge.
+        t2: Second `CTerm` to merge.
+
+    Returns:
+        A `CTerm` which is the merge of the two input `CTerm` instances.
+    """
+    # check all cells in t1 and t1, if they are the same, keep them, otherwise, create a new free variable for them
+    t1_config, t1_subst = split_config_from(t1.config)
+    t2_config, t2_subst = split_config_from(t2.config)
+
+    if t1_config != t2_config:
+        # cannot merge two configurations with different structure
+        return None
+
+    new_subst = Subst({})
+    new_t1_subst = Subst({})
+    new_t2_subst = Subst({})
+
+    for cell in t1_subst:
+        if t1_subst[cell] == t2_subst[cell] and not var_occurrences(t1_subst[cell]):
+            # keep the cell if it is the same & exists no free variables
+            new_subst = new_subst * Subst({cell: t1_subst[cell]})
+        else:
+            # create a new free variable for the cell
+            new_t1_subst = new_t1_subst * Subst({cell: t1_subst[cell]})
+            new_t2_subst = new_t2_subst * Subst({cell: t2_subst[cell]})
+    new_config = new_subst(t1_config)
+
+    new_constraints: list[KInner] = []
+    for new_subst, t in [(new_t1_subst, t1), (new_t2_subst, t2)]:
+        if new_subst:
+            antecedent = mlAnd([mlEquals(KVariable(cell), new_subst[cell]) for cell in new_subst])
```

Comment:
I think this can be simplified as well? See https://github.com/runtimeverification/k/blob/master/pyk/src/pyk/kast/inner.py#L753

```suggestion
            antecedent = new_subst.ml_pred
```

Reply:
Thank you for point it out! I've changed that!